### ISSUE TITLE
Add validation when a gRPC service is configured on a HTTP/1 cluster

### ISSUE THREAD

**Issue by mpuncel** [envoyproxy/envoy#11439]
When we were prototyping using the ext_authz filter, we configured it to use gRPC to a "self cluster", i.e. one that points at a socket on the local Envoy sidecar. This is perhaps non-standard, but the reason we did this is that we wanted the requests to run through our typical egress filter chain.

We forgot to set http2_protocol_options on this cluster, and were quite confused when our gRPC upstream service was returning OK status codes, and we were still seeing requests get denied by the filter. What happened was the filter was making HTTP/1 requests to the self cluster, and then they were upgraded to http2 before being sent to the gRPC upstream. The response then had the gRPC trailers stripped, so the filter failed the requests.

The clue to what was happening was that our egress logs were showing HTTP/1 rather than HTTP/2, everything else including gRPC trailers in the access log looked fine!

Sample log: ( the `-/0` is the grpc-status header followed by grpc-status trailer)
```
[2020-02-27T00:27:45.898Z] "POST /envoy.service.auth.v2.Authorization/Check HTTP/1.1" 200 -/0 - 2979 11 57 57 "-" "797296e2-5515-428a-945f-3459265494fa" "multipass_safe--staging" "-" "-" "10.1.3.17:32508"
```

This is probably an obscure enough issue to be low priority, but I have 2 possible suggestions:

1) validate that the filter can only be configured with an h2 capable cluster when using gRPC (although this raises questions about when the validation should happen)
2) force gRPC client requests to h2 even when the cluster is not configured for h2

**Comment by mattklein123** [comment]
cc @snowp @alyssawilk this is related to common ALPN/protocol misconfiguration in the cluster config, related to work you have both been doing recently. Anything we can do here to make this better would be great.

**Comment by cthumuluru** [comment]
+1; I had the similar issue.

**Comment by snowp** [comment]
I think I prefer the second option outlined, since iirc there aren't any cases where you wouldn't use HTTP/2 for gRPC. 

Would we eventually want to support HTTP/3? If so, validating that its either HTTP/2 or HTTP/3 might be be best option.

**Comment by cthumuluru** [comment]
@snowp — I would like to work on this issue; I too prefer the second option outline in the issue.

**Comment by dio** [comment]
@cthumuluru do you have something on this already? WIP or anything? I can assign this issue to you.

**Comment by cthumuluru** [comment]
@dio, I've not started on this yet. I'll start looking at it this week and provide an update next week.

**Comment by alyssawilk** [comment]
Hm, this makes me wonder if we should have required that one of 
explicit | downstream | alpn be configured when using the new protocol options.
I think it's new enough we could force one of them to be set.  It wouldn't necessarily prevent misconfiguration but it would at least make it visible at a glance that the cluster had http/1 config vs http/2 config.
@mattklein123 @htuch  think it's worth it, or should we just do the grpc => H2 checks and call it a day?

**Comment by htuch** [comment]
It would definitely be clearer, would this be a breaking change though?

**Comment by mattklein123** [comment]
> It would definitely be clearer, would this be a breaking change though?

Yeah it sounds like a breaking change to me. Often gRPC would be configured as prior knowledge H2 and plenty of time without TLS/ALPN, so it seems like just checking for H2 would be great?

**Comment by alyssawilk** [comment]
It would be a breaking change on config which landed 7 days ago: https://github.com/envoyproxy/envoy/pull/14079

I'm good either way, making it explicit (now, before folks pick up the new config) or someone can just do the H2 checks later on.

**Comment by htuch** [comment]
It's a good point that this landed only recently. For security issues we have a defined window where we just fix on master. @envoyproxy/api-shepherds should we add some similar grace period for breaking changes or is this going to be hard for consumers?

**Comment by markdroth** [comment]
I'd be in favor of a small grace period, like maybe a couple of weeks.  It seems like within that window, anyone who has started using the new fields will still have enough context to make whatever changes are needed fairly easily.  And I'd rather do that than have to deprecate something that we just recently added.

**Comment by mattklein123** [comment]
Sorry I just read this thread again and I didn't fully understand before what was being proposed.'

> I think it's new enough we could force one of them to be set. It wouldn't necessarily prevent misconfiguration but it would at least make it visible at a glance that the cluster had http/1 config vs http/2 config.

I think I suggested this during the original code review so +1 from me.

**Comment by htuch** [comment]
I'm good with this, but we should clarify this and make it explicit in https://github.com/envoyproxy/envoy/blob/master/api/API_VERSIONING.md. I'll take it offline with API shepherds.
